mixer/pkg/runtime/testing/data/data.go

```diff
@@ -467,6 +485,57 @@ spec:
     - ihalt1.thalt.istio-system
 `
 
+// RuleCheckOutput1 is a standard testing rule for check output template
+var RuleCheckOutput1 = `
+apiVersion: config.istio.io/v1alpha2
+kind: rule
+metadata:
+  name: rcheckoutput1
+  namespace: istio-system
+spec:
+  actions:
+  - handler: hcheckoutput1.acheckoutput
+    instances:
+    - icheckoutput1.tcheckoutput.istio-system
+    name: sample
+  requestHeaderOperations:
+  - name: '"user"'
+    values:
+    - sample.output.value
+`
+
+// RuleCheckOutput2 is a testing rule for check output template with multiple outputs
+var RuleCheckOutput2 = `
+apiVersion: config.istio.io/v1alpha2
+kind: rule
+metadata:
+  name: rcheckoutput2
+  namespace: istio-system
+spec:
+  actions:
+  - handler: hcheckoutput1.acheckoutput
+    instances:
+    - icheckoutput1.tcheckoutput.istio-system
+    name: a
+  - handler: hcheckoutput1.acheckoutput
+    instances:
+    - icheckoutput1.tcheckoutput.istio-system
+    name: b
+  requestHeaderOperations:
+  - name: '"a"'
+    values:
+    - a.output.value
+    operation: REPLACE
+  - name: '"user"'
+    operation: REMOVE
+  responseHeaderOperations:
+  - name: '"b"'
+    values:
+    - b.output.value
+    - prefix.generated.string
+    operation: APPEND
```

Comment:
I battled so much with myself to make peace with these operations at the same level as actions... In the end, after a very long while thinking about it I managed to accept the design. Just to be sure, all instances handled by each handle in each action need to pass their Check with status 200 in order to all of the operations to be executed, right? And the outputs are built based on the actionName . output . outputKey

Reply:
Yes, that's right. Action name is scoped within the rule, output is a fixed symbol that references the output of the action, and outputKey is adapter-specific field in the output schema.

We discussed this quite a bit within the group (@geeknoid @mandarjog @douglas-reid ), and we felt this strikes the balance of:
- disallowing adapters from performing unchecked mutations on the data path
- supporting generic type system for the adapters, adapters should be able to define arbitrary output templates
- aligns well with existing Attribute-producing adapters, whereas the operator controls how output flows back into mixer's data path